### PR TITLE
[core] Pick node from top k

### LINKED ISSUE #26262: [core] More than 30s scheduling delay in embarrassingly parallel task workload after scaling from 100 to 300 cores
### What happened + What you expected to happen

Hi Ray team, I’m experiencing some scalability issue when I’m testing my cluster with more nodes. Here’s a bit of context.

    ray version: 1.9
    workload: 300 tasks each of which takes about a minute to complete in a single core
    cluster: around 300 cores running on a k8s platform. Each node is configured with 15 cores. Before I run my test, all nodes are brought up and all 300 cores are available.
    execution pattern: I use a very simple parallel map pattern that sends out 300 tasks at once and wait for them all.

    performance
        100 cores:
            this is the case that time can add up. For example, 300 tasks (each task takes about 1 min on a single core) in total take less than 3 min to finish.
            in addition, from Ray Dashboard, I see all nodes are taken up quickly.
        300 cores:
            when I increase my cores to 300, I’m expecting the total amount of time is around 1 min. However, it takes more than 2 min to finish.
            In this case, from Ray Dashboard, I see tasks are sent to worker much slower than the case of 100 cores and at most it takes around 200 cores at the same time.
            I also tried getting timeline stat by setting RAY_PROFILINEG=1 and the timeline tells a similar story as what I observed from dashboard

            Notice that some workers are scheduled much much later than others, though I send all tasks at the same time and the cluster has enough cores.

I also tried “gang scheduling” following instructions from this page using placement group (I’m using ray 1.9) : [Placement Group Examples — Ray v1.9.2](https://docs.ray.io/en/releases-1.9.2/auto_examples/placement-group.html?highlight=scheduling#)
but it doesn’t make much difference.

We found this issue when we did our final regression test before first time go-live with Ray, and it’s now a blocker. I would be really appreciate if any one could provide any suggestion. I’m glad to provide any further information if needed.

### Versions / Dependencies

1.9, need reproduction on 1.13 and master.

### Reproduction script

```python
def price(...):
    ...
    pricing_tasks = list(
        map(lambda batch: price_remote.remote(batch), pricing_batches)
    )
    return ray.get(pricing_tasks)
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

> Today with the default scheduling policy, Ray will try to pack tasks on nodes until the resource utilization is beyond a certain threshold and spread tasks afterwards.
This has caused slow down the scheduling speed for embarrassingly parallel jobs: 

- we will only move on to another node before the current node's resource if sufficiently utilized,
- for each node, the overhead of accepting new job and starting of a new workers is not negligible.    
- the overall scheduling speed doesn't scale with the number of nodes; 

This PR is one proposal to address the problem: instead of stick to one node, we randomly choose one node from top-k nodes for the default scheduling, where the node is sorted by it's resource utilization in reverse order. 

Intuitively, this allows us to kick off the workers startup on multiple node in parallel of the scheduling.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stephanie-wang** [comment]
No improvement on many_tasks nightly test unfortunately, but I'm pretty sure this PR + a fix for #28252 will do it. Closing for now while that issue is being resolved.

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc [hunk 1]
```diff
@@ -23,117 +23,140 @@ namespace ray {
...
+  if (top_k <= 0) {
+    top_k = std::max(
+        static_cast<uint64_t>(RayConfig::instance().max_pending_lease_requests_per_scheduling_category()),
+        static_cast<uint64_t>(RayConfig::instance().scheduler_top_k_fraction() * nodes_.size()));
+  }
+  if (!available_nodes.empty()) {
+    // First prioritize available nodes.
+    return GetBestNode(available_nodes,
+                       !force_spillback && local_node_is_available,
+                       1,
```
Comment (jjyao) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc:150]: 1 -> top_k

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc [hunk 2]
```diff
@@ -23,117 +23,140 @@ namespace ray {
...
+                       spread_threshold);
+  } else if (!feasible_nodes.empty() && !require_node_available) {
+    // If there are no available nodes, and the caller is okay with an
+    // unavailable node, check the feasible nodes next.
+    return GetBestNode(
+        feasible_nodes, !force_spillback && local_node_is_feasible, 1, spread_threshold);
```
Comment (jjyao) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc:156]: 1 -> top_k

**Review by jjyao — COMMENTED** [review]
After fixing the bug, top_k works as expected. For many_tasks, num_tasks_per_second increased from 26 to 514.

**Comment by wumuzi520** [comment]
> @wumuzi520 is this the top-k strategy you implemented internally? Could you comment on the benefits you have seen?

Yes, we already use the top-k strategy internally.

I think the advantage of the random strategy in top-k is that when there are many tasks, it is possible for several nodes with high scores to initialize the environment in parallel, without being constrained by the pack-50% limit of a single node under hybrid scheduling. The disadvantage is that due to the introduction of random, the autoscaler may be affected to a certain extent, but it is generally acceptable.

**Review by ericl — COMMENTED** [review]
Does this also work for the SPREAD policy?

**Comment by scv119** [comment]
> Does this also work for the SPREAD policy?
I wonder how SPREAD help out the embarrassingly parallel job as well. Let me run the experiment as well.

**Comment by scv119** [comment]
```
342     0.05GiB /home/ray/anaconda3/bin/python -m ray.util.client.server --address=172.31.217.235:9031 --host=0.0.0.
Success! Started 10000 tasks in 324.2963795661926s. (411.5839552455203 tasks/s)
```

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.h [hunk 3]
```diff
@@ -49,24 +53,17 @@ class HybridSchedulingPolicy : public ISchedulingPolicy {
  public:
   HybridSchedulingPolicy(scheduling::NodeID local_node_id,
                          const absl::flat_hash_map<scheduling::NodeID, Node> &nodes,
-                         std::function<bool(scheduling::NodeID)> is_node_available)
+                         std::function<bool(scheduling::NodeID)> is_node_alive)
```
Comment (scv119) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.h:56]: TODO: update comments.

**Comment by rkooo567** [comment]
Is there any benchmark result?

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.h [hunk 4]
```diff
@@ -78,18 +75,48 @@ class HybridSchedulingPolicy : public ISchedulingPolicy {
     kNonGpu
   };
 
+  bool IsNodeFeasible(const scheduling::NodeID &node_id,
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.h:78]: can you add docstrings?

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.h [hunk 5]
```diff
@@ -78,18 +75,48 @@ class HybridSchedulingPolicy : public ISchedulingPolicy {
...
+                      const NodeFilter &node_filter,
+                      const NodeResources &node_resources,
+                      const ResourceRequest &resource_request) const;
+
+  float CompuateLocalNodeScore(float spread_threshold) const;
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.h:83]: same for functions below

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc [hunk 6]
```diff
@@ -23,146 +23,188 @@ namespace ray {
...
+        // It's okay if the local node's pull manager is at
+        // capacity because we will eventually spill the task
+        // back from the waiting queue if its args cannot be
+        // pulled.
+        ignore_pull_manager_at_capacity = true;
+        local_node_is_feasible = true;
+      }
+      bool is_available =
+          node_resources.IsAvailable(resource_request, ignore_pull_manager_at_capacity);
+      if (node_id == local_node_id_ and is_available) {
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc:129]:
```suggestion
      if (node_id == local_node_id_&& is_available) {
```

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc [hunk 7]
```diff
@@ -23,146 +23,188 @@ namespace ray {
...
+                        static_cast<int32_t>(nodes_.size() * scheduler_top_k_fraction)));
+
+  if (!available_nodes.empty()) {
+    bool prioritize_local_node = !force_spillback && local_node_is_available;
+    // First prioritize available nodes.
+    return GetBestNode(
+        available_nodes,
+        num_candidate_nodes,
+        prioritize_local_node,
+        prioritize_local_node ? CompuateLocalNodeScore(spread_threshold) : 1);
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc:158]:
```suggestion
        /* local_node_score*/prioritize_local_node ? CompuateLocalNodeScore(spread_threshold) : 1);
```

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.h [hunk 8]
```diff
@@ -78,18 +75,48 @@ class HybridSchedulingPolicy : public ISchedulingPolicy {
...
   /// nodes), kNonGpu (can only schedule on non-GPU nodes.
   ///
   /// \return -1 if the task is unfeasible, otherwise the node id (key in `nodes`) to
   /// schedule on.
-  scheduling::NodeID HybridPolicyWithFilter(const ResourceRequest &resource_request,
-                                            float spread_threshold,
-                                            bool force_spillback,
-                                            bool require_available,
-                                            NodeFilter node_filter = NodeFilter::kAny);
+  scheduling::NodeID ScheduleImpl(const ResourceRequest &resource_request,
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.h:98]: Looks like the docstring is incorrect?

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc [hunk 9]
```diff
@@ -23,146 +23,188 @@ namespace ray {
...
+  } else if (!feasible_nodes.empty() && !require_node_available) {
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc:159]: what is `require_node_available` for?

Comment (rkooo567) [reply src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc:159]: can you add comment on this bool why this condition means to "prioritize_local_node"?

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc [hunk 10]
```diff
@@ -23,146 +23,188 @@ namespace ray {
...
+  RAY_CHECK(num_candidate_nodes >= 1);
+  // Pick the top num_candidate_nodes nodes with the lowest score.
+  // First, sort nodes so that we always break ties between nodes in the same
+  // order.
+  std::sort(
+      node_scores.begin(),
+      node_scores.end(),
+      [](const std::pair<scheduling::NodeID, float> &a,
+         const std::pair<scheduling::NodeID, float> &b) { return a.first < b.first; });
+  std::stable_sort(
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc:79]: `// Second, sort by the score`

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc [hunk 11]
```diff
@@ -23,146 +23,188 @@ namespace ray {
...
+      node_scores.begin(),
+      node_scores.end(),
+      [](const std::pair<scheduling::NodeID, float> &a,
+         const std::pair<scheduling::NodeID, float> &b) { return a.second < b.second; });
+
+  if (prioritize_local_node) {
+    if (local_node_score <= node_scores.front().second) {
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc:86]: is this condition necessary? Maybe we can just return `local_node_id_; `?

**Comment by scv119** [comment]
to avoid confusion, all follow up development moved to here: https://github.com/ray-project/ray/pull/31868

**Pull request closed without merging**